### PR TITLE
[Bugfix] Fix of crash in TCanvas as suggested by Olivier Couet

### PR DESCRIPTION
**Pull request by andriish**
This fixes a crash in the following code:

```
#include <TCanvas.h>
#include <TText.h>
#include <TPad.h>
#include <fenv.h>

int main(int argc, char **argv) {
    TCanvas* C= new TCanvas("canvas","canvas",1024,768);
    C->cd();
    TPad* c= new TPad();
    //c->cd(1);//works
    c->cd();//fails
    TText* t= new TText(0.5,0.5,"some text");
    t->Draw();
    delete t;
    delete c;
    delete C;
    return 0;
}
``` 
that shoud be compiled to an executable.
Tag @couet

### CODE REVIEW

**Review by couet — APPROVED** [review]

**Comment by couet** [comment]
@andriish does this change fixes the problem in your main program ?

File: graf2d/gpad/src/TCanvas.cxx [hunk 1]
```diff
@@ -779,7 +779,7 @@ void TCanvas::Close(Option_t *option)
 {
    TPad    *padsave = (TPad*)gPad;
    TCanvas *cansave = 0;
-   if (padsave) cansave = (TCanvas*)gPad->GetCanvas();
+   if (padsave && gPad->TestBit(kNotDeleted)) cansave = (TCanvas*)gPad->GetCanvas();
```
Comment (pcanal) [inline graf2d/gpad/src/TCanvas.cxx:782]: This is sub-optiomal has this test (TestBit(kNoDeleted)) can be foiled if the memory re-used between the deletion and the execution.  So question "why" in this case is gPad still pointing to a deleted object?

**Review by pcanal — CHANGES_REQUESTED** [review]
We need better understanding of the failure mode.

**Comment by andriish** [comment]
Hi @couet , 

First of all thanks for the fix and thanks for asking!

> does this change fixes the problem in your main program 

Yes and no. It improves a lot. 

I hope with some more fixes there will be no issues on the ROOT side.

But, of course, there ~~might be~~  are some "native" bugs, but these I will have to fix by myself.

Best regards,

Andrii

P.S. Right now the only nontrivial (i.e. non FPE ) nonreported issue I see is a crash related to TIter. But I cannot reproduce it standalone...

**Comment by andriish** [comment]
>
Hi @pcanal ,
> We need better understanding of the failure mode.
OK, I see.

Meanwhile I've simplified the reproducer
```
#include <TCanvas.h>
#include <TPad.h>
int main(int argc, char **argv) {
    TCanvas* C= new TCanvas("canvas","canvas",1024,768);
    C->cd();
    TPad* c= new TPad();
    c->cd();
    delete c;
    delete C;
    return 0;
}
```


+1 −1
Conversation 2
Commits 1
Checks 0
Files changed 1
Conversation
@andriish
Contributor
andriish commented 19 minutes ago

This fixes a crash in the following code:

#include <TCanvas.h>
#include <TText.h>
#include <TPad.h>
#include <fenv.h>

int main(int argc, char **argv) {
    TCanvas* C= new TCanvas("canvas","canvas",1024,768);
    C->cd();
    TPad* c= new TPad();
    //c->cd(1);//works
    c->cd();//fails
    TText* t= new TText(0.5,0.5,"some text");
    t->Draw();
    delete t;
    delete c;
    delete C;
    return 0;
}

that shoud be compiled to an executable.
Tag @couet
Andrii Verbytskyi
Fix of crash in TCanvas as suggested by Olivier Couet
c4d690d
@andriish andriish requested a review from couet as a code owner 19 minutes ago
@phsft-bot
Collaborator
phsft-bot commented 19 minutes ago

Can one of the admins verify this patch?
@couet
couet approved these changes 17 minutes ago
@couet couet requested a review from pcanal 16 minutes ago
@couet couet self-assigned this 16 minutes ago
@couet
Member
couet commented 15 minutes ago

@andriish does this change fixes the problem in your main program ?
@pcanal
pcanal reviewed 13 minutes ago
graf2d/gpad/src/TCanvas.cxx
@@ -779,7 +779,7 @@ void TCanvas::Close(Option_t *option)
{
   TPad    *padsave = (TPad*)gPad;
   TCanvas *cansave = 0;
   if (padsave) cansave = (TCanvas*)gPad->GetCanvas();
   if (padsave && gPad->TestBit(kNotDeleted)) cansave = (TCanvas*)gPad->GetCanvas();
@pcanal
pcanal 13 minutes ago Member

This is sub-optiomal has this test (TestBit(kNoDeleted)) can be foiled if the memory re-used between the deletion and the execution. So question "why" in this case is gPad still pointing to a deleted object?
@andriish
@pcanal
pcanal requested changes 13 minutes ago
Member
pcanal left a comment

We need better understanding of the failure mode.
@andriish
Contributor Author
andriish commented 7 minutes ago

Hi @couet ,

First of all thanks for the fix and thanks for asking!

    does this change fixes the problem in your main program

Yes and no. It improves a lot.

I hope with some more fixes there will be no issues on the ROOT side.

But, of course, there might be are some "native" bugs, but these I will have to fix by myself.

Best regards,

Andrii

File: graf2d/gpad/src/TCanvas.cxx (hunk 1 above)
Comment (couet) [reply graf2d/gpad/src/TCanvas.cxx:782]: the pad "c" is deleted explicitly by the macro pad4.C and deleted again when the canvas C is closed. gPad was set to "c" and was not set to null when c  was deleted (I guess). So a better change might be to set gPad to null when a pad equal to gPad is deleted. Let me see if a such mechanism can be done.

Comment (andriish) [reply graf2d/gpad/src/TCanvas.cxx:782]: ``if (this==gPad) gPad=nullptr;``  in the ``~TPad`` ?
or somathing like 

    ``if (this==gPad) gPad=fMother;``  in the also in  ``~TPad``

Comment (couet) [reply graf2d/gpad/src/TCanvas.cxx:782]: I do not know. testing now.

Comment (couet) [reply graf2d/gpad/src/TCanvas.cxx:782]: If I undo the change in `TCanvas::Close()`  and change the `~TPad()` the following way:
```
TPad::~TPad()
{
   if (!TestBit(kNotDeleted)) return;
   if (gPad == this) gPad = nullptr;  // <<<< new test

```

Then it also fixes the problem. ie: pad4.cxx works and stressGraphics also.

Comment (andriish) [reply graf2d/gpad/src/TCanvas.cxx:782]: Wouldnt fMother make more sense?

Comment (couet) [reply graf2d/gpad/src/TCanvas.cxx:782]: It sounds more logical to test `gPad` seems to me. The ctor delete `this` pad and if `this` is `gPad`, then `gPad` must be null also.  I am not sure to understand the `fMother` logic.

Comment (andriish) [reply graf2d/gpad/src/TCanvas.cxx:782]: I think that is just a convention. 
One can say the logic is that after the deletion of a subpad, gPad points to the parent pad. But null is another option.

Comment (couet) [reply graf2d/gpad/src/TCanvas.cxx:782]: I guess we can leave it like that for the time being.

**Comment by couet** [comment]
@andriish so now this PR should be modified to suppress changes in TCanvas and change TPad destructor to reset gPad if needed. Can you do it ? (that's your PR).

**Comment by andriish** [comment]
Hi @couet  ,

I've put a line to the ~TPad(). It resets gPad to nullptr. 

Best regards,
Andrii

**Review by couet — APPROVED** [review]
It is ok.
Thanks

**Comment by andriish** [comment]
Hi @couet ,

I will test the nightbuild and close the relevant issues tomorrow.

Best regards,

Andrii

**Pull request merged**